File: libAssertions/src/main/java/ua/com/dekalo/assertions/Assertions.java
/*
 * Copyright (c) 2017 Dekalo Stanislav. All rights reserved.
 */
package ua.com.dekalo.assertions;

import android.os.Handler;
import android.os.Looper;

import java.util.Map;

/**
 * Main purpose is to add various assertions in IS_PRODUCTION mode.
 * And switch it off in production.
 * <p/>
 * Created by dekalo on 25.08.15.
 */
public final class Assertions {

    private static final CrashReporter EMPTY_REPORTER = new CrashReporter() {
        @Override
        public void report(Throwable throwable) {
        }
    };

    private static boolean CRASH_ON_FAIL = false;
    private static Handler UI_HANDLER = new Handler(Looper.getMainLooper());
    private static CrashReporter CRASH_REPORTER = EMPTY_REPORTER;

    public interface CrashReporter {
        void report(Throwable throwable);
    }

    private static class ThrowDelegateRunnable implements Runnable {

        private final RuntimeException runtimeException;

        private ThrowDelegateRunnable(Throwable throwable) {
            this.runtimeException = new RuntimeException(throwable);
        }

        @Override
        public void run() {
            throw runtimeException;
        }
    }

    public static void setCrashReporter(CrashReporter crashReporter) {
        CRASH_REPORTER = crashReporter == null ? EMPTY_REPORTER : crashReporter;
    }

    public static boolean isCrashOnFailEnabled() {
        return CRASH_ON_FAIL;
    }

    public static void setCrashOnFail(boolean enabled) {
        CRASH_ON_FAIL = enabled;
    }

    private static void fail(final Throwable e, boolean silently) {
        CRASH_REPORTER.report(e);
        if (CRASH_ON_FAIL && !silently) {
            UI_HANDLER.post(new ThrowDelegateRunnable(e));
        }
    }

    /**
     * Assertion failed with throwable.
     * <p>
     * Same as fail(...) method, but it will not crash application even if  setCrashOnFail(true) was called.
     * <p>
     * This is useful for triggering possible but not desired condition in client code.
     */
    public static void failSilently(final Throwable throwable) {
        fail(throwable, true);
    }

    /**
     * Assertion failed with throwable.
     */
    public static void fail(final Throwable throwable) {
        fail(throwable, false);
    }

    /**
     * Raise exception from ExceptionFactory.
     */
    public static void fail(final ExceptionFactory exceptionFactory) {
        fail(exceptionFactory.create());
    }

    /**
     * Ensures that method called on UI thread, else raises exception provided by ExceptionFactory.
     */
    public static void checkUIThread(ExceptionFactory exceptionFactory) {
        if (Looper.getMainLooper() != Looper.myLooper()) {
            fail(exceptionFactory);
        }
    }

    /**
     * Checks that shouldBeTrue condition is true, else raises exception provided by ExceptionFactory.
     */
    public static void assertTrue(boolean shouldBeTrue, ExceptionFactory exceptionFactory) {
        if (!shouldBeTrue) {
            fail(exceptionFactory);
        }
    }

    /**
     * Checks that shouldBeFalse condition is false, else raises exception provided by ExceptionFactory.
     */
    public static void assertFalse(boolean shouldBeFalse, ExceptionFactory exceptionFactory) {
        if (shouldBeFalse) {
            fail(exceptionFactory);
        }
    }

    /**
     * Checks that size of enum equal to size of enum mapping, else raises exception provided by ExceptionFactory.
     */
    public static void assertFullMapping(Class<? extends Enum> enumType, Map enumMapping, ExceptionFactory factory) {
        assertTrue(enumType.getEnumConstants().length == enumMapping.size(), factory);
    }
}

File: libAssertions/src/main/java/ua/com/dekalo/assertions/Assertions.kt
/*
 * Copyright (c) 2017 Dekalo Stanislav. All rights reserved.
 */
package ua.com.dekalo.assertions

import java.util.*


/**
 * Main purpose is to add various assertions in IS_PRODUCTION mode.
 * And switch it off in production.
 *
 * Assertion not works until you will add AssertionHandler.
 *
 * Created by dekalo on 25.08.15.
 */
object Assertions {

    private const val DEFAULT_PRIOTIY = 0;

    private val handlers = TreeSet<Pair<Int, AssertionHandler>>()

    /**
     * Add AssertionHandler to handler list.
     */
    @JvmOverloads
    fun addAssertionHandler(assertionHandler: AssertionHandler, priority: Int = DEFAULT_PRIOTIY) {
        handlers.add(Pair(priority, assertionHandler))
    }

    /**
     * Remove AssertionHandler from handlers list.
     */
    fun removeAssertionHandler(assertionHandler: AssertionHandler) {
        while (handlers.remove(handlers.find { it.second == assertionHandler })) {
            // all logic already in while statement
        }
    }

    private fun hasHandlers(): Boolean = handlers.isNotEmpty()

    /**
     * Assertion failed with throwable.
     *
     * @param silently will be passed to AssertionHandler, crash will not happen in any case if it is silent assertion. It is reasonable if it is possible but undesired situation that you would like to log in your crash reporting tool.
     */
    @JvmOverloads
    fun fail(throwable: Throwable, silently: Boolean = false) {
        if (hasHandlers()) {
            handlers.forEach { it.second.report(throwable, silently) }
        }
    }

    internal fun fail(exceptionFactory: ExceptionFactory) {
        if (hasHandlers()) {
            exceptionFactory.create().let { throwable ->
                handlers.forEach { it.second.report(throwable, false) }
            }
        }
    }

    /**
     * Checks that shouldBeTrue condition is true, else raises exception provided by ExceptionFactory.
     */
    fun assertTrue(shouldBeTrue: Boolean, exceptionFactory: ExceptionFactory) {
        if (!shouldBeTrue) {
            fail(exceptionFactory)
        }
    }

    /**
     * Checks that shouldBeFalse condition is false, else raises exception provided by ExceptionFactory.
     */
    fun assertFalse(shouldBeFalse: Boolean, exceptionFactory: ExceptionFactory) {
        if (shouldBeFalse) {
            fail(exceptionFactory)
        }
    }

    fun assertNull(shouldBeNull: Any?, exceptionFactory: ExceptionFactory) {
        if (shouldBeNull != null) {
            fail(exceptionFactory)
        }
    }

    fun assertNotNull(shouldNotBeNull: Any?, exceptionFactory: ExceptionFactory) {
        if (shouldNotBeNull == null) {
            fail(exceptionFactory)
        }
    }

    fun <T> assertEmpty(shouldBeEmpty: Collection<T>, exceptionFactory: ExceptionFactory) {
        if (shouldBeEmpty.isNotEmpty()) {
            fail(exceptionFactory)
        }
    }

    fun <T> assertNotEmpty(shouldNotBeEmpty: Collection<T>, exceptionFactory: ExceptionFactory) {
        if (shouldNotBeEmpty.isEmpty()) {
            fail(exceptionFactory)
        }
    }
}
